### DISCUSSION TITLE
Consider deprecating the `draw_<layout>` functions?

### DISCUSSION THREAD

**Discussion by rossbar** [networkx/networkx#5220] [Q&A] [answered]
There are currently several convenience functions included in the `nx_pylab` module that allow users to draw a graph with a non-spring layout without having to explicitly provide one.

In practice, these are thin wrappers around `nx.draw`, replacing the specification of the layout via the `pos` kwarg with a new function name, i.e. `nx.draw_circular(G)` is equivalent to `nx.draw(G, pos=nx.circular_layout(G))`.

IMO there are several downsides to these convenience functions:
 - They are not well-documented. The docstring for each essentially says "draw G with <whichever layout> - see `draw_networkx` for details". This leads to issues like #5215 
 - They increase the API surface, which isn't a huge deal, but does lead to duplication. For example, solving the above bullet point would likely involve duplicating information that is already in the layout docstrings.
 - They're out-of-sync with the layouts functions, i.e. there are several more layout functions available that don't have a corresponding `nx.draw_<layout>`  - see e.g. `nx.bipartite_layout`, `nx.multipartite_layout`, and `nx.spiral_layout`. It'd be easy to add the "missing" `draw_<layout>` functions, but it'd be another thing to keep in sync whenever a new layout function is added (*though in principle this could be done automatically*).
 - This usage pattern masks configurability of the layouts: many of the layout functions have options that can tweak the behavior, but these options are not forwarded to the layout functions when calling `draw_<layout>`
 - Related to the above, this usage pattern is less performant and also masks the *reusability* of layouts. When using `draw_<layout>`, a new layout (i.e. dict mapping nodes to 2 or 3D coordinates) is computed internally for each call. For computationally expensive layout functions, this can be a significant performance hit - see e.g. stef4k/nx-guides#2. I would argue that computing a layout once and reusing it for multiple drawing calls should generally be preferred.
 - The `draw_<layout>` functions fall afoul of the "There should be one, and preferably only one, obvious way to do it" principle

Arguments in favor of the convenience functions:
 - They are indeed *convenient* after all, especially for quick exploratory visualization where users want to get a sense of a graph without diving too deeply into layout configurations.
 - They arguably improve discoverability of the different layouts. For example, if a user types `nx.draw<tab>` in IPython or a notebook, they will see the `draw_<layout>` functions.
 - Counterpoint from PEP 20: "practicality beats purity"

Given the above list of pros/cons, I'd be inclined to deprecate the `nx.draw_<layout>` functions in favor of the `nx.draw(G, pos=nx.<layout>)` pattern. One concern is that it might be a noisy deprecation (I have no idea how frequently these functions are used) but I think this is somewhat mitigated by 1) a major release, 2) the ability to have informative error messages that explicitly state how to fix the issue, and 3) the ability to add the functions to a module `__get_attr__` to continue raising informative instructions if users try to access `draw_<layout>` functions even in NetworkX 3.0.

I'm curious to hear what others think, and please don't hesitate to share additional pros/cons that I've missed!

**Comment by MridulS** [comment, marked as answer]
I'm -0.5 on removing these functions as they are *extremely* convenient.

> They arguably improve discoverability of the different layouts. For example, if a user types nx.draw<tab> in IPython or a notebook, they will see the draw_<layout> functions.

This is pretty much how I quickly draw out the graph if `nx.draw` gives a hairball. The thinking being that there should be atleast one layout where the output isn't a hairball. Writing down `nx.draw_` and then hitting `tab` a couple of times to find a good layout is a quick process because of these convenient functions.

As you have stated that there are indeed a *lot* of downsides of keeping these quick layout convenience functions, I think we should try to fix them and improve documentation rather than deprecating these functions.

> They are not well-documented. The docstring for each essentially says "draw G with - see draw_networkx for details". This leads to issues like Documentation for draw_shell does not explain all arguments #5215

Agree, we should update the documentation.

> They increase the API surface, which isn't a huge deal, but does lead to duplication. For example, solving the above bullet point would likely involve duplicating information that is already in the layout docstrings.

I think this is something that's hard to avoid.

> They're out-of-sync with the layouts functions, i.e. there are several more layout functions available that don't have a corresponding nx.draw_<layout> - see e.g. nx.bipartite_layout, nx.multipartite_layout, and nx.spiral_layout. It'd be easy to add the "missing" draw_<layout> functions, but it'd be another thing to keep in sync whenever a new layout function is added (though in principle this could be done automatically).

Agree, another issue that we should fix.

> This usage pattern masks configurability of the layouts: many of the layout functions have options that can tweak the behavior, but these options are not forwarded to the layout functions when calling draw_<layout>

Obviously I don't know how other folks use these functions in their workflow but calls to `draw_<layout>` are usually during the initial exploratory part, the "production" code moves to `nx.draw` as it exposes/documents it in a neater way (?). Or maybe I just came up with this system as `draw_<layout>` functions are indeed not properly setup with the underlying layout functions.

> Related to the above, this usage pattern is less performant and also masks the reusability of layouts. When using draw_<layout>, a new layout (i.e. dict mapping nodes to 2 or 3D coordinates) is computed internally for each call. For computationally expensive layout functions, this can be a significant performance hit - see e.g. Suggestion: Reuse layouts for plotting stef4k/nx-guides#2. I would argue that computing a layout once and reusing it for multiple drawing calls should generally be preferred.

Agree, maybe we should look at some sort of local cache solutions for this?

> The draw_<layout> functions fall afoul of the "There should be one, and preferably only one, obvious way to do it" principle

I rationalise this to myself with the argument that `draw_<layout>` is just there for exploration/testing/convenience. The "production" workflow will use `nx.draw` with all the arguments and layout designs. So I see `draw_<layout>`  as a different thing from `nx.draw`.

Let me know if this makes any sense 😅

It would be great if others could also weigh in on this!

**Comment by rossbar** [comment]
Thanks for weighing in @MridulS - you make excellent points. I hadn't considered the split between "exploratory" and "production" use-cases --- looking at it from this perspective I think you've made a very strong case for having multiple interfaces to drawing with different layouts. I also agree that many of the issues I raised in the OP can be resolved by updating documentation.

In summary, I agree with you: thanks for the insights! Let's table the deprecation idea and instead work on updating the docs and making sure all the layouts that we want to conveniently expose are accounted for. However, if in the future anyone wants to weigh in please feel free to comment!
